Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/VirtIOGPU/VirtIOGPU.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
+namespace Kernel::Graphics {
+
+VirtIOGPU::VirtIOGPU(PCI::Address address)
+    : VirtIODevice(address, "VirtIOGPU")
+    , m_scratch_space(MM.allocate_contiguous_kernel_region(8 * PAGE_SIZE, "VirtGPU Scratch Space", Region::Access::Read | Region::Access::Write))
+{
+    VERIFY(!!m_scratch_space);
+    dbgln("Gonna configure now");
+    if (auto cfg = get_config(ConfigurationType::Device)) {
+        m_device_configuration = cfg;
+        bool success = negotiate_features([&](u64 supported_features) {
+            u64 negotiated = 0;
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_VIRGL))
+                dbgln("VirtIOGPU: VIRGL is not yet supported!");
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_EDID))
+                dbgln("VirtIOGPU: EDID is not yet supported!");
+            return negotiated;
+        });
+        if (success) {
+            read_config_atomic([&]() {
+                m_num_scanouts = config_read32(*cfg, 0x8);
+            });
+            dbgln("VirtIOGPU: num_scanouts: {}", m_num_scanouts);
+            success = setup_queues(2); // CONTROLQ + CURSORQ
+        }
+        if (success) {
+            finish_init();
+        }
+        // 1. Get display information using VIRTIO_GPU_CMD_GET_DISPLAY_INFO
+        query_display_information();
+        // 2. Create BUFFER using VIRTIO_GPU_CMD_RESOURCE_CREATE_2D
+        create_framebuffer();
+        // 3. Attach backing storage using  VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING
+        attach_backing_storage();
+        // 4. Use VIRTIO_GPU_CMD_SET_SCANOUT to link the framebuffer to a display scanout.
+        attach_framebuffer_to_scanout();
+        // 5. Render our test pattern
+        draw_test_pattern();
+        // 6. Use VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D to update the host resource from guest memory.
+        transfer_framebuffer_data_to_host(m_display_info.rect);
+        // 7. Use VIRTIO_GPU_CMD_RESOURCE_FLUSH to flush the updated resource to the display.
+        flush_displayed_image(m_display_info.rect);
+    } else {
+        VERIFY_NOT_REACHED();
```

Comment:
should we really verify not reached here?

Reply:
I'm not entirely sure what we should do here then, the graphics management subsystem tried to fail. I think in this case the best course of action would be to panic.

Reply:
i would expect the graphics management subsystem to fall back to text mode, like it would if virtio gpu support wasnt available at all

Reply:
Hmm, I guess that would be possible. I'm not sure if it is possible to go back from virtio mode to text mode, but I'll try to make it fall back to text mode.

TBH for virtiogpu it might be better to just VERIFY fail in that case, if it doesn't initialize the framebuffer the user would probably either need to switch to a different backend or fix their qemu setup 99% of the time.

Reply:
hmm, since its virtio thats probably fine then.